### DISCUSSION TITLE
Upgrading Dependencies on Windows

### DISCUSSION THREAD

**Discussion by OldGrumpyViking** [juftin/hatch-pip-compile#59] [Ideas]
The scripts provided in the [readme.md](https://github.com/juftin/hatch-pip-compile?tab=readme-ov-file#upgrading-dependencies) for upgrading dependencies works in linux only as setting an environment variable in windows for a command is not done in the same way.

Now i have tried to make a seperate script for linux and for windows without success, example:
```bash
upgrade-all = "PIP_COMPILE_UPGRADE=1 hatch env run --env {env_name} -- python --version"
upgrade-all-win = "$env:PIP_COMPILE_UPGRADE=1 ; hatch env run --env {env_name} -- python --version"
```

My current working workaround is to use `hatch shell` followed by `$env:PIP_COMPILE_UPGRADE=1 ; pip -V`

Any ideas?

The best fix i can think of to not have linux and windows specific commands is to introduce a command to the hatch cli from the plugin, but this feature is not yet possible but there is a PR ongoing https://github.com/pypa/hatch/issues/798
Then something like `hatch pip-compile upgrade` would be possible.

**Comment by oprypin** [comment]
I think hatch has big limitations regarding user-configured commands. It runs commands just with the default shell  which can be wildly different on Windows, meaning that you're forced to use only "polyglot" commands - no variables of any kind, no wildcards, and also using only double quotes.

I think your best bet at the moment is to not add this to Hatch (too self-referential anyway?) and keep those commands only in your shell history. Also as I keep saying, the most ergonomic way to perform an upgrade is just to delete the lock file 😅

**Comment by juftin** [comment]
So I've had an idea for a `hatch-pip-compile` CLI for a little while now that would take care of this for you. I decided to code it up today and would love your thoughts @OldGrumpyViking @oprypin ([see code here](https://github.com/juftin/hatch-pip-compile/compare/feat/cli)). It's an extra tool that users would need to learn (blegh, I know). It will eventually be unnecessary once https://github.com/pypa/hatch/issues/798 / https://github.com/pypa/hatch/pull/1094 are accepted and hatch has a way to do this itself.

This would be a CLI outside of hatch so you would need to install it separately - but essentially all it does is temporarily set the `PIP_COMPILE_UPGRADE` / `PIP_COMPILE_UPGRADE_PACKAGES` environment variables for you and invoke hatch for a given environment. It uses the same options as the regular `pip-compile` CLI.

## Examples

```shell
pipx install hatch-pip-compile
```

### `--upgrade` the `default` environment

The below command will upgrade all packages in the `default` environment.

```shell
hatch-pip-compile --upgrade default
```

### Upgrade a specific package

The below command will upgrade the `requests` package in the `default`environment.

```shell
hatch-pip-compile --upgrade-package requests default
```

### Upgrade all `pip-compile` environments

The below command will upgrade all packages in all `pip-compile` environments.

```shell
hatch-pip-compile --upgrade --all
```

**Comment by OldGrumpyViking** [comment]
This looks very promising to me. It would be nice if the cli is compatible to the envisioned hatch plugin cli such that the only change for a future version would be to remove the "-" between hatch and pip-compile.

**Comment by juftin** [comment]
Following up here, the `hatch-pip-compile` CLI is now released on `1.10.0` - hopefully this is an ergonomic options on all platforms. @OldGrumpyViking let me know how you like it on Windows https://github.com/juftin/hatch-pip-compile?tab=readme-ov-file#using-the-hatch-pip-compile-cli

**Comment by OldGrumpyViking** [comment]
@juftin this is really nice and useful!

The link at in the [README.md](https://github.com/juftin/hatch-pip-compile/blob/main/README.md?plain=1#L90) is wrong so it took me a while to find 🙈. Seems to be due to the refactor commit 1d3dd127ee538dba4a66647d35516fec3d8a5a45.

I seem to be able to avoid having to install the plugin in the root python installation by prepending hatch run to the command:
``` pyproject.toml
[tool.hatch.envs.default.scripts]
upgrade-all = "hatch env run --env {env_name} -- hatch-pip-compile -U"
upgrade-pkg = "hatch env run --env {env_name} -- hatch-pip-compile -P '{args}'"
```
`hatch run upgrade-all`

The upgrade all works as intended!

However when using `hatch run upgrade-pkg pydantic` it states that it will update, but the package is not updated. I have not had the chance to look further into this.
```bash
hatch-pip-compile: Upgrading packages: 'pydantic'
hatch-pip-compile: Targeting environments: default
hatch-pip-compile: Running `hatch env run --env default -- python --version`
```
